File: step_15_graphs/15.1_learning/P004_dfs_traversal/Solution.java
/**
 * Problem: DFS Traversal
 * Difficulty: EASY | XP: 10
 *
 * Implement DFS on a graph. Return nodes in DFS order.
 * Handle disconnected graphs.
 *
 * @author DSA_Nova
 */

import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: RECURSIVE DFS
    // Time: O(V + E)  |  Space: O(V)
    // ============================================================
    public static List<Integer> dfsRecursive(int V, List<List<Integer>> adj) {
        boolean[] visited = new boolean[V];
        List<Integer> result = new ArrayList<>();

        // Outer loop handles disconnected components
        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                dfsHelper(i, adj, visited, result);
            }
        }
        return result;
    }

    private static void dfsHelper(int node, List<List<Integer>> adj,
                                   boolean[] visited, List<Integer> result) {
        visited[node] = true;
        result.add(node);

        for (int neighbor : adj.get(node)) {
            if (!visited[neighbor]) {
                dfsHelper(neighbor, adj, visited, result);
            }
        }
    }

    // ============================================================
    // APPROACH 2: ITERATIVE DFS (Explicit Stack)
    // Time: O(V + E)  |  Space: O(V)
    // ============================================================
    public static List<Integer> dfsIterative(int V, List<List<Integer>> adj) {
        boolean[] visited = new boolean[V];
        List<Integer> result = new ArrayList<>();

        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                Deque<Integer> stack = new ArrayDeque<>();
                stack.push(i);

                while (!stack.isEmpty()) {
                    int node = stack.pop();
                    if (visited[node]) continue;

                    visited[node] = true;
                    result.add(node);

                    // Push neighbors in reverse order so that the smallest
                    // neighbor is processed first (matching recursive behavior)
                    List<Integer> neighbors = adj.get(node);
                    for (int j = neighbors.size() - 1; j >= 0; j--) {
                        int neighbor = neighbors.get(j);
                        if (!visited[neighbor]) {
                            stack.push(neighbor);
                        }
                    }
                }
            }
        }
        return result;
    }

    // ============================================================
    // HELPER: Build adjacency list from edge pairs
    // ============================================================
    private static List<List<Integer>> buildGraph(int V, int[][] edges) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]); // undirected
        }
        return adj;
    }

    public static void main(String[] args) {
        System.out.println("=== DFS Traversal ===\n");

        // Test 1: Connected graph
        //   0 --- 1
        //   |     |
        //   2     4
        int[][] edges1 = {{0,1}, {0,2}, {1,4}};
        List<List<Integer>> adj1 = buildGraph(5, edges1);
        System.out.println("Test 1 (connected):");
        System.out.println("  Recursive:  " + dfsRecursive(5, adj1));
        adj1 = buildGraph(5, edges1); // rebuild for fresh test
        System.out.println("  Iterative:  " + dfsIterative(5, adj1));

        // Test 2: Disconnected graph
        //   0 --- 1       3
        //   |     |       |
        //   2     4       5
        int[][] edges2 = {{0,1}, {0,2}, {1,4}, {3,5}};
        List<List<Integer>> adj2 = buildGraph(6, edges2);
        System.out.println("\nTest 2 (disconnected):");
        System.out.println("  Recursive:  " + dfsRecursive(6, adj2));
        adj2 = buildGraph(6, edges2);
        System.out.println("  Iterative:  " + dfsIterative(6, adj2));

        // Test 3: Single node
        List<List<Integer>> adj3 = buildGraph(1, new int[][]{});
        System.out.println("\nTest 3 (single node):");
        System.out.println("  Recursive:  " + dfsRecursive(1, adj3));

        // Test 4: No edges (all disconnected)
        List<List<Integer>> adj4 = buildGraph(4, new int[][]{});
        System.out.println("\nTest 4 (no edges):");
        System.out.println("  Recursive:  " + dfsRecursive(4, adj4));
    }
}

File: step_15_graphs/15.1_learning/P004_dfs_traversal/solution.py
"""
Problem: DFS Traversal
Difficulty: EASY | XP: 10

Implement DFS on a graph. Return nodes in DFS order.
Handle disconnected graphs.
"""
from typing import List
from collections import defaultdict


# ============================================================
# APPROACH 1: RECURSIVE DFS
# Time: O(V + E)  |  Space: O(V)
# ============================================================
def dfs_recursive(V: int, adj: List[List[int]]) -> List[int]:
    """
    Standard recursive DFS with outer loop for disconnected components.
    """
    visited = [False] * V
    result = []

    def dfs(node: int):
        visited[node] = True
        result.append(node)
        for neighbor in adj[node]:
            if not visited[neighbor]:
                dfs(neighbor)

    # Outer loop: handles disconnected components
    for i in range(V):
        if not visited[i]:
            dfs(i)

    return result


# ============================================================
# APPROACH 2: ITERATIVE DFS (Explicit Stack)
# Time: O(V + E)  |  Space: O(V)
# ============================================================
def dfs_iterative(V: int, adj: List[List[int]]) -> List[int]:
    """
    Iterative DFS using explicit stack. Avoids recursion depth issues.
    """
    visited = [False] * V
    result = []

    for i in range(V):
        if not visited[i]:
            stack = [i]
            while stack:
                node = stack.pop()
                if visited[node]:
                    continue
                visited[node] = True
                result.append(node)

                # Push in reverse order so smallest neighbor is popped first
                for neighbor in reversed(adj[node]):
                    if not visited[neighbor]:
                        stack.append(neighbor)

    return result


# ============================================================
# HELPER: Build adjacency list
# ============================================================
def build_graph(V: int, edges: List[List[int]]) -> List[List[int]]:
    adj = [[] for _ in range(V)]
    for u, v in edges:
        adj[u].append(v)
        adj[v].append(u)
    return adj


if __name__ == "__main__":
    print("=== DFS Traversal ===\n")

    # Test 1: Connected graph
    #   0 --- 1
    #   |     |
    #   2     4
    adj1 = build_graph(5, [[0,1], [0,2], [1,4]])
    print("Test 1 (connected):")
    print(f"  Recursive:  {dfs_recursive(5, adj1)}")
    adj1 = build_graph(5, [[0,1], [0,2], [1,4]])
    print(f"  Iterative:  {dfs_iterative(5, adj1)}")

    # Test 2: Disconnected graph
    #   0 --- 1       3
    #   |     |       |
    #   2     4       5
    adj2 = build_graph(6, [[0,1], [0,2], [1,4], [3,5]])
    print("\nTest 2 (disconnected):")
    print(f"  Recursive:  {dfs_recursive(6, adj2)}")
    adj2 = build_graph(6, [[0,1], [0,2], [1,4], [3,5]])
    print(f"  Iterative:  {dfs_iterative(6, adj2)}")

    # Test 3: Single node
    adj3 = build_graph(1, [])
    print(f"\nTest 3 (single node): {dfs_recursive(1, adj3)}")

    # Test 4: No edges
    adj4 = build_graph(4, [])
    print(f"Test 4 (no edges):    {dfs_recursive(4, adj4)}")
